pkg/kube/kube_handler.go

```diff
@@ -100,6 +119,12 @@ func NewClusterConfig(inClusterConfig *rest.Config, userToken string, cluster st
 	config := rest.CopyConfig(inClusterConfig)
 	config.BearerToken = userToken
 	config.BearerTokenFile = ""
+	// Modify the default number of requests that the given client can do
+	// This is useful to handle a large number of namespaces which are check in parallel
+	// Burst is the initial number of request made in parallel
+	// Then further requests are performed following the QPS rate
+	config.Burst = maxReq * 2
```

Comment:
Does this mean we need to give the var a better name? If "Burst is the initial number of requests made in parallel" then it doesn't make sense to me to make it twice the `maxReq` value. Perhaps `queriesPerSecond`?

Reply:
This is a typical tocket bucket algorithm. According to the k8s docs: 
_The rate limiter allows bursts of up to 'burst' to exceed the QPS, while still maintaining a smoothed qps rate of 'qps'. The bucket is initially filled with 'burst' tokens, and refills at a rate of 'qps'. The maximum number of tokens in the bucket is capped at 'burst'._

So, technically, in a single instant, you can perform up to `burst` requests (that is, it is the `maxReq` indeed).  However, I don't think you can make `burst` requests in the first second. Instead, you will have to wait `burst/qps`  seconds until you are able to perform `burst` requests.

As a result, `burst/qps` would be the time to wait until you can perform another spike of `burst ` requests. That is, by setting this value to `2`, we are allowing to exceed the `qps` each 2 seconds. So a user who has set the "maximum requests" will see that their value is being "ignored".

Here https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/ they use burst=30,qps=20, that is, 1.5s between spikes. However, the spike is just +50% of the qps, not the +100% as we are configuring here.

So, since it is a configuration that really depends upon the cluster hardware, I'd opt for making the qps also configurable and defaulting both params to (burst=30,qps=20).
I think a production-ready cluster would benchmark how much time do the requests take and, given that value, they would set the time between spikes.

Reply:
> However, I don't think you can make `burst` requests in the first second. Instead, you will have to wait `burst/qps` seconds until you are able to perform `burst` requests.

Why? The docs, as you mentioned, say that you can "while still maintaining a smoothed qps rate of 'qps'". If you could only make `burst` requests in `burst/qps` seconds, then you'd not be exceeding the QPS at all, where as it explicitly says you can. My interpretation was that as long as the smoothed QPS rate matches the qps setting (say over 5 s - just as an example), then you can do a burst that exceeds `qps` requests in the first second as long as qps matches over 5 seconds (again, just an example, in reality it's probably based on the gradient getting back towards the set qps).

> As a result, `burst/qps` would be the time to wait until you can perform another spike of `burst ` requests. That is, by setting this value to `2`, we are allowing to exceed the `qps` each 2 seconds. So a user who has set the "maximum requests" will see that their value is being "ignored".

Right, this would be the result if there was no smoothing: if burst was 200, qps was 100, you could do an initial burst of 200 requests, but the qps wouldn't allow you to add any more until 2s. But my assumption is with smoothing this would allow further requests as long as the smoothed qps is going back down to the expected?

> So, since it is a configuration that really depends upon the cluster hardware, I'd opt for making the qps also configurable and defaulting both params to (burst=30,qps=20).

+1 to make both configurable, if Andres agrees (I don't feel strongly - would also be happy with 1.5*qps for burst), but regarding defaults, 30/20 is what `kube-controller-manager` uses as defaults (important part of the k8s control plane). Not sure we should use the same... I'd leave the defaults as they are [in the client-go code (DefaultQPS: 5, DefaultBurst: 10)](https://github.com/kubernetes/client-go/blob/master/rest/config.go#L44), as long as users can configure.

Reply:
Note that now (10qps/15burst), for 200 namespaces, it takes 19s to load... quite a lot.

Reply:
> Why? The docs, as you mentioned, say that you can "while still maintaining a smoothed qps rate of 'qps'". If you could only make `burst` requests in `burst/qps` seconds, then you'd not be exceeding the QPS at all, where as it explicitly says you can. My interpretation was that as long as the smoothed QPS rate matches the qps setting (say over 5 s - just as an example), then you can do a burst that exceeds `qps` requests in the first second as long as qps matches over 5 seconds (again, just an example, in reality it's probably based on the gradient getting back towards the set qps).

It is a very interesting topic, I haven't checked the actual implementation, but I have assumed something like:
![image](https://user-images.githubusercontent.com/11535726/112018259-b1a3f180-8b2e-11eb-87c6-10dfa49cfc74.png)
(image credits to Roman Voloboev)

Indeed, see an implementation from the k8s go client:
```go
// NewTokenBucketRateLimiter creates a rate limiter which implements a token bucket approach.
// The rate limiter allows bursts of up to 'burst' to exceed the QPS, while still maintaining a
// smoothed qps rate of 'qps'.
// The bucket is initially filled with 'burst' tokens, and refills at a rate of 'qps'.
// The maximum number of tokens in the bucket is capped at 'burst'.

```
https://github.com/kubernetes/client-go/blob/f6ce18ae578c8cca64d14ab9687824d9e1305a67/util/flowcontrol/throttle.go#L48-L52


> I can leave the default as qps: 10, burst: 15. I think that leaving numbers below that really hurts the UX for larger clusters.

I'm ok with these values as well

Reply:
The graph isn't for a smoothed/rolling 5s window though, hence why you get spikes? I agree that if it's not maintaining a smoothed qps rate, but rather a fixed intervals (non-smoothed) then it would be spiked as you say - I'd just understood (without reading the implementation) that it was a smoothed qps rate in this case. But yep, interesting - I wish we did more basic DS and Algo discussions and learning :)

I'm also fine with the values. Thanks!